Yank/analyze.py

```diff
@@ -32,27 +35,1145 @@
+        _, nunsampled, _ = unsampled_energy_matrix.shape
+        # Subsample data to obtain uncorrelated samples
+        N_k = np.zeros(nstates, np.int32)
+        N = len(niterations)  # number of uncorrelated samples
+        N_k[:] = N
+        mbar_ready_energy_matrix = sampled_energy_matrix
+        if nunsampled > 0:
+            fully_interacting_u_ln = unsampled_energy_matrix[:, 0, :]
+            noninteracting_u_ln = unsampled_energy_matrix[:, 1, :]
+            # Augment u_kln to accept the new state
+            new_energy_matrix = np.zeros([nstates + 2, nstates + 2, N], np.float64)
+            N_k_new = np.zeros(nstates + 2, np.int32)
+            # Insert energies
+            new_energy_matrix[1:-1, 0, :] = fully_interacting_u_ln
+            new_energy_matrix[1:-1, -1, :] = noninteracting_u_ln
+            # Fill in other energies
+            new_energy_matrix[1:-1, 1:-1, :] = sampled_energy_matrix
+            N_k_new[1:-1] = N_k
+            # Notify users
+            logger.info("Found expanded cutoff states in the energies!")
+            logger.info("Free energies will be reported relative to them instead!")
+            # Reset values, last step in case something went wrong so we dont overwrite u_kln on accident
+            mbar_ready_energy_matrix = new_energy_matrix
+            N_k = N_k_new
+        return mbar_ready_energy_matrix, N_k
+
+    def _compute_free_energy(self):
+        """
+        Estimate free energies of all alchemical states.
+        """
+
+        # Create MBAR object if not provided
+        if self._mbar is None:
+            self._create_mbar_from_scratch()
+
+        nstates = self.mbar.N_k.size
+
+        # Get matrix of dimensionless free energy differences and uncertainty estimate.
+        logger.info("Computing covariance matrix...")
+
+        try:
+            # pymbar 2
+            (Deltaf_ij, dDeltaf_ij) = self.mbar.getFreeEnergyDifferences()
+        except ValueError:
+            # pymbar 3
+            (Deltaf_ij, dDeltaf_ij, theta_ij) = self.mbar.getFreeEnergyDifferences()
+
+        # Matrix of free energy differences
+        logger.info("Deltaf_ij:")
+        for i in range(nstates):
+            str_row = ""
+            for j in range(nstates):
+                str_row += "{:8.3f}".format(Deltaf_ij[i, j])
+            logger.info(str_row)
+
+        # Matrix of uncertainties in free energy difference (expectations standard
+        # deviations of the estimator about the true free energy)
+        logger.info("dDeltaf_ij:")
+        for i in range(nstates):
+            str_row = ""
+            for j in range(nstates):
+                str_row += "{:8.3f}".format(dDeltaf_ij[i, j])
+            logger.info(str_row)
+
+        # Return free energy differences and an estimate of the covariance.
+        free_energy_dict = {'value': Deltaf_ij, 'error': dDeltaf_ij}
+        self._computed_observables['free_energy'] = free_energy_dict
+
+    def get_free_energy(self):
```

Comment:
I like the idea of having both a `.free_energy` and a `.free_energies`, but I don't know how best make that work with the `ObservablesRegistry`